### PR TITLE
[Incremental, Frontend, NFC] Redo interface to `ReferencedNameTracker`

### PR DESCRIPTION
**Pull request by davidungar**
Cleans up interface to `ReferencedNameTracker` to allow for use information for cross-type dependencies.

### CODE REVIEW

**Comment by davidungar** [comment]
@swift-ci please smoke test

File: include/swift/AST/ReferencedNameTracker.h [hunk 1]
```diff
@@ -14,31 +14,103 @@
...
+  llvm::DenseMap<NameAndUse, bool> DynamicLookupNames;
+  llvm::DenseMap<HolderMemberAndUse, bool> UsedMembers;
+
+public:
+  /// Record a top level use.
+  /// \param name The name being used
+  /// \param isCascadingUse Whether the dependency from use to def transitively affects
+  /// uses of the use.
+  /// \param use The declaration using the name, if known.
+  void addTopLevelName(DeclBaseName name, bool isCascadingUse,
```
Comment (CodaFi) [inline include/swift/AST/ReferencedNameTracker.h]: From what I can see, no callers actually pass a third parameter here.

**Comment by CodaFi** [comment]
I think this should be done in two steps:
- In true LLVM style, we should make a separate "referenced name tracker 2.0" for the cross-type dependencies
- Then we should consider the old thing legacy code. 

Ideally this should just stamp out the macros from the legacy interface.

**Comment by futurejones** [comment]
Can one of the admins verify this patch?

**Pull request closed without merging**